[FSDP] Add exec order validation

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#75902 [FSDP] Add exec order validation**

This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

Commit:
[FSDP] Add exec order validation

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.



[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

awgu commented:
Failures are real. Working on it...

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1803,6 +1853,10 @@ def _pre_backward_hook(*unused: Any) -> None:
                 # pre-backward hook.
                 torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
 
+            # Start of a backward pass for the first time in an backward pass.
```
I moved this state transition to earlier in the pre-backward hook since otherwise the call to `_rebuild_full_params()` is still under `IDLE` rather than `PRE_BACKWARD`. The training state should not affect the execution of any of the code between this new position and the old position a few lines down. However, it could be good to get a second pair of eyes to sanity check this.

rohan-varma replied:
Hmmm, I guess it depends on whether we want to consider this all_gather as part of backward_pre or not?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+            index = torch.tensor([param_index], dtype=torch.int32).to(device)
+            dist._all_gather_base(indices, index, group=self.process_group)
+            # Check that all ranks plan to all-gather the same parameter index
+            for (r1, i1), (r2, i2) in itertools.combinations(
```
I am doing this pairwise check so that the error message can concretely include where execution orders differ (i.e. which rank and which index).

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+        execution order differs from that of the first iteration, meaning that
+        we can no longer guarantee correct execution.
+
+        Executing in ``no_sync()`` does not affect the execution order check
```
The previous commit in the stack was important to ensuring this.

awgu commented on test/distributed/fsdp/test_fsdp_exec_order.py:
```diff
@@ -0,0 +1,145 @@
+        inp = fsdp_model.module.get_input(self.device)
+        with context:
+            fsdp_model(*inp)
+
```
I did not include "valid" tests since the existing test suite are all "valid". I did not see any false positive warnings when running through `test_fsdp_core.py`, but please feel free to let me know if after this is landed, we start seeing warnings for other tests.

awgu reviewed (commented):
Some comments to help review

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+            eod.index += 1
+        else:
+            device = param.device
+            indices = torch.zeros(self.world_size, dtype=torch.int32).to(device)
```
I think this incurs extra CPU -> GPU copy, can we do torch.zeros(world_size, dtype=torch.int32, device=device)

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+            device = param.device
+            indices = torch.zeros(self.world_size, dtype=torch.int32).to(device)
+            index = torch.tensor([param_index], dtype=torch.int32).to(device)
+            dist._all_gather_base(indices, index, group=self.process_group)
```
what happens if user is gathering the same parameter index but the module somehow has different shape? Does broadcast at the beginning or torch_distributed_debug anticipated to help with this?

awgu replied:
I agree then it is problematic. I guess with your diff for checking parameter shapes, this is not an issue since the shapes will already be validated.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+            for (r1, i1), (r2, i2) in itertools.combinations(
+                ((rank, indices[rank]) for rank in range(self.world_size)), 2,
+            ):
+                if torch.equal(i1, i2):
```
can't we just do `if not torch.equal(...)` and raise the error directly?

awgu replied:
Yes. I included to the `continue` so that the error message has one fewer indent, but I can error directly if that is preferred.

rohan-varma commented on test/distributed/fsdp/test_fsdp_exec_order.py:
```diff
@@ -0,0 +1,145 @@
+            output = fsdp_model(*inp)
+            loss = fsdp_model.module.get_loss(inp, output).to(self.device)
+            fsdp_model.module.run_backward(loss)
+        context = self.assertWarns(expected_warning=UserWarning) \
```
Wonder if the warning will be hit every single iteration, will this become a bit too verbose for users who change the module in the same way across all ranks?

awgu replied:
Changing this to warn only once per rank.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -232,6 +233,35 @@ class OptimStateKeyType(Enum):
+        """Called in :meth:`_wait_for_post_backward` to reset data for the next
+        iteration."""
+        self.is_first_iter = False
+        self.index = 0
```
It looks like `param_order` also gets modified, do we not need to reset this?

awgu replied:
Yes, `param_to_param_index` is initialized once and never changed, and param_order` is constructed over the course of the first iteration and not changed thereafter.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -622,6 +655,19 @@ def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        assert self._is_root, "Expects to be called only on root FSDP modules"
+        param_to_param_index = {
+            cast(FlatParameter, param): i
+            for i, param in enumerate(self.parameters())
```
Does `.parameters()` only contain `FlatParameter` in all cases? Is there a reason we can't use `self.params`?

awgu replied:
`self.params` only contains the root module's parameters. `.parameters()` includes the `FlatParameter`s of all nested FSDP modules as well.

rohan-varma reviewed (approved):
LGTM, just some minor questions / concerns for your consideration. @zhaojuanmao can also take a look, thank you!

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -232,6 +233,35 @@ class OptimStateKeyType(Enum):
+    This contains the data used for validating execution order across ranks.
+
+    Attributes:
+        param_to_param_index (Dict[FlatParameter, int]): Mapping from FSDP
```
Yeah. This is just simply enumerating the flattened parameters. Another approach would have been to `enumerate(self.parameters())` every time, but I decided to save it here.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2277,6 +2333,7 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
                         # Allocate based on full size from all shards.
                         _alloc_storage(p._full_param_padded, size=p_full_size)  # type: ignore[attr-defined]
                         output_tensor = p._full_param_padded  # type: ignore[attr-defined]
+                    self._check_all_gather(p)
```
Could you clarify the confusion with the name? I meant that it is checking the all-gather in the next line, not that the check itself runs an all-gather.

awgu commented on test/distributed/fsdp/test_fsdp_exec_order.py:
```diff
@@ -0,0 +1,145 @@
+        if self.rank != 0:
+            fsdp_model.flip_path()
+        inp = fsdp_model.module.get_input(self.device)
+        with context:
```
Yes, this will be a false positive. However, since the warning message is written to say that a change has been detected, not necessarily that something is wrong, hopefully it is not too confusing for the user.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+            # Warn possibly multiple times for the same iteration since the
+            # number of warnings may be informative in debugging
+            if eod.index >= len(eod.param_order) and self.rank == 0:
+                print("eod.index:", eod.index)
+                print("eod.param_order:", eod.param_order)
```
could we give Torch_Warning_Once()?

zhaojuanmao replied:
also does this need to return here? otherwise it will crash if it calls eod.param_order[eod.index] in the next line?

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+                print("eod.param_order:", eod.param_order)
+            if param_index != eod.param_order[eod.index]:
+                warnings.warn(
+                    "Execution order differs from that of the first iteration "
```
Torch_Warning_Once?

zhaojuanmao reviewed (approved):
overall looks great! some minor comments. Like how it handles mismatch before hang!

zhaojuanmao commented:
forgot to mention that we should run benchmarks before landing this PR, as it touched core codes

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+        if self.training_state not in (
+            TrainingState_.FORWARD, TrainingState_.BACKWARD_PRE,
+        ):
+            return
```
This `_check_all_gather()` prefaces the all-gather in `_rebuild_full_params()`. However, we may call `_rebuild_full_params()` in training states other than `FORWARD` and `BACKWARD_PRE`, in which case we do not want to include this check. (For example, we call `_rebuild_full_params()` in `summon_full_params()`.)

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+                    continue
+                raise RuntimeError(
+                    "Forward pass execution order differs across ranks: rank "
+                    f"{r1} is all-gathering parameter {i1} while rank {r2} "
```
awesome!

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

Commit:
Update on "[FSDP] Add exec order validation"


This adds execution order validation for `FSDP`.

**Background**
For collectives, we may classify three types of mismatches:
1. Collective type mismatch: Rank `i` schedules one collective type while rank `j` schedules another collective type.
	- This errors immediately.
2. Collective shape mismatch: Ranks schedule the same collective but one rank's contribution has a mismatched shape.
	- This errors immediately.
3. Logical collective mismatch: Collective type and shapes match but rank `i`'s collective should not logically match with rank `j`'s collective.
	- This error silently.

Our goal with this execution order validation is to improve the user experience around class-3 mismatches. `TORCH_DISTRIBUTED_DEBUG=DETAIL` may help with class-1 and class-2 mismatches (not implemented yet).

**Approach**
Assumption 1: The `FSDP` model `model` is wrapped identically across ranks. This allows each rank to establish a common `param` to `param_index` bijection (by enumerating `model.parameters()`).
Assumption 2: If the backward pass all-gathers are in order, then the backward pass reduce-scatters are in order. Hence, this approach only checks all-gathers.

- First iteration:
    - Before every all-gather (in `_rebuild_full_params()`), we all-gather a one-dimension `torch.tensor(param_index)` from each rank indicating the parameter the rank is about to all-gather. All `param_index` s should be the same.
- Subsequent iterations:
	- We record the parameter order. If the order deviates from that of the first iteration, then we issue a *warning* to the user.

Differential Revision: [D35684277](https://our.internmc.facebook.com/intern/diff/D35684277)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35684277).

awgu commented:
I updated the internal diff with the relevant benchmark flows. To be fully transparent, I do see around 1-2% P50 QPS regression in the 1 node 8 GPU setting (consistently across 3 runs for both RoBERTa and ResNet).

I also made some non-trivial code changes (though the overall logic is still generally the same), so I am re-requesting review.

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2331,6 +2437,97 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
+                else:
+                    msg_suffix = f"the FSDP module wrapping {param_names}"
+                sub_msg = msg_prefix + msg_suffix
+                warnings.warn(
```
nit: do we want to warn only once at the first time when the mismatch is found? because the main purpose is for users to notice this change, possibly one warning is sufficient?

awgu replied:
From earlier offline discussion in the devs chat, I think it is still good to include warnings through the first deviating iteration in case the order change is somehow unintentional. Then, providing some additional information may be helpful for debugging.
I think it should be fine since the messages are still contained to be for one iteration and will not flood the console constantly.

awgu commented:
@pytorchbot merge this please
